src/proc/occlusion-filter.cpp

```diff
@@ -45,51 +81,116 @@ namespace librealsense
+    void occlusion_filter::monotonic_heuristic_invalidation(float3* points, float2* uv_map, const std::vector<float2>& pix_coord, const rs2::depth_frame& depth) const
     {
         float occZTh = 0.1f; //meters
         int occDilationSz = 1;
-        auto pixels_ptr = pix_coord.data();
         auto points_width = _depth_intrinsics->width;
         auto points_height = _depth_intrinsics->height;
+        auto pixels_ptr = pix_coord.data();
+        auto points_ptr = points;
+        auto uv_map_ptr = uv_map;
+        float maxInLine = -1;
+        float maxZ = 0;
 
-        for (size_t y = 0; y < points_height; ++y)
+        auto frame_size = static_cast<int>(_depth_intrinsics->width * _depth_intrinsics->height);
+        std::allocator<byte> alloc;
+        byte* depth_planes[1];
+
+        if (_occlusion_scanning == horizontal)
         {
-            float maxInLine = -1;
-            float maxZ = 0;
-            int occDilationLeft = 0;
 
-            for (size_t x = 0; x < points_width; ++x)
             {
-                if (points->z)
+                scoped_timer t1("Horizontal Scan");
+                for (size_t y = 0; y < points_height; ++y)
                 {
-                    // Occlusion detection
-                    if (pixels_ptr->x < maxInLine || (pixels_ptr->x == maxInLine && (points->z - maxZ) > occZTh))
+
+                    maxInLine = -1;
+                    maxZ = 0;
+                    int occDilationLeft = 0;
+
+                    for (size_t x = 0; x < points_width; ++x)
                     {
-                        uv_map->x = 0.f;
-                        uv_map->y = 0.f;
-                        occDilationLeft = occDilationSz;
+                        if (points_ptr->z)
+                        {
+                            //Occlusion detection
+                            if (pixels_ptr->x < maxInLine || (pixels_ptr->x == maxInLine && (points_ptr->z - maxZ) > occZTh))
+                            {
+                                *points_ptr = { 0, 0, 0 };
+                                occDilationLeft = occDilationSz;
+                            }
+                            else
+                            {
+                                maxInLine = pixels_ptr->x;
+                                maxZ = points_ptr->z;
+                                if (occDilationLeft > 0)
+                                {
+                                    *points_ptr = { 0, 0, 0 };
+                                    occDilationLeft--;
+                                }
+                            }
+                        }
+                        ++points_ptr;
+                        ++uv_map_ptr;
+                        ++pixels_ptr;
                     }
-                    else
+                }
+            }
+        }
+        else if (_occlusion_scanning == vertical)
+        {
+            depth_planes[0] = (byte*)alloc.allocate(depth.get_bytes_per_pixel() * frame_size);
```

Comment:
Let's discuss how this can be reduced. Allocations are also performance sensitive

Reply:
I don't have another alternative currently